server/controllers/admin/roles.js

```diff
@@ -0,0 +1,211 @@
+const db = require('../../lib/db');
+const q = require('q');
+const _ = require('lodash');
+const FilterParser = require('../../lib/filter');
+
+module.exports.list = list;
+module.exports.detail = detail;
+module.exports.create = create;
+module.exports.update = update;
+module.exports.remove = remove;
+module.exports.affectPages = affectPages;
+module.exports.affectToUser = affectToUser;
+module.exports.listForUser = listForUser;
+
+function list(req, res, next) {
+
+  const filters = new FilterParser(req.query, { tableAlias : 'r' });
+
+  let sql = `
+    SELECT BUID(r.uuid) as uuid, r.label, r.project_id 
+    FROM role  r
+  `;
+
+  filters.equals('project_id');
+  sql = filters.applyQuery(sql);
+
+  db.exec(sql, filters.parameters())
+    .then((rows) => {
+      res.json(rows);
+    }).catch(next)
+    .done();
+}
+
+
+function detail(req, res, next) {
+  const sql = `
+   SELECT BUID(uuid) as uuid, label, project_id 
+   FROM role
+   WHERE  uuid = ?  LIMIT 1
+  `;
+  const binaryUuid = db.bid(req.params.uuid);
+  db.one(sql, binaryUuid)
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+// create a new role
+function create(req, res, next) {
+  var sql = `
+    INSERT INTO  role(uuid, label, project_id)
+    VALUES(?, ?,?)
+  `;
+
+  db.exec(sql, [db.uuid(), req.body.label, req.body.project_id])
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+function update(req, res, next) {
+  const role = req.body;
+  delete role.uuid;
+  const sql = ` UPDATE role SET ? WHERE uuid = ?`;
+
+  db.exec(sql, [role, db.bid(req.params.uuid)])
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+
+function remove(req, res, next) {
+  const binaryUuid = db.bid(req.params.uuid);
+
+  const sql = `DELETE FROM role WHERE uuid = ?`;
+  db.exec(sql, binaryUuid)
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+// affect permission to a specific role
+
+function affectPages(req, res, next) {
+  const data = req.body;
+
+  const unitIds = data.unit_ids;
+  const roleUuid = db.bid(data.role_uuid);
+
+  const msg = {
+    state : 1,
+    message : 'success',
+  };
+  const deleteFromRole = `DELETE FROM role_unit WHERE role_uuid = ? `;
+  const affectPage = `
+    INSERT INTO  role_unit(uuid, unit_id, role_uuid)
+    VALUES( ?, ?, ?)`;
+
+  db.exec(deleteFromRole, roleUuid)
+    .then(() => {
+      const t = [];
+
+      if (!unitIds) {
+        return;
+      }
+      if (_.isArray(unitIds)) {
+        unitIds.forEach((id, idx) => {
```

Comment:
Instead of using an index to make a new array, just use [Array.prototype.map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map).  It will return a new array of whatever return values you specify.  Like this:
```js
const promises = unitIds.map(id => db.exec(affectPage, [db.uuid(), id, roleUuid]));

return q.all(promises);
```